keps/sig-node/3063-dynamic-resource-allocation/README.md

```diff
@@ -0,0 +1,1763 @@
+  to control.
+
+  For example, a single resource driver might manage different kinds of
+  FPGAs. For each kind, one ResourceClass could define the kind and provide paths
+  to kind-specific tools or URLs for additional resources. Only a handful of
+  ResourceClasses are expected per resource driver.
+
+- ResourceClaim, namespaced, with parameters provided by a normal user
+  that describes a resource instance that needs to be allocated. A
+  ResourceClaim contains the usual metadata, a spec and a status. The
+  spec identifies the driver that handles the resource via a class
+  name.
+
+A third one with information about the driver ("ResourceDriver", similar to the
+"CSIDriver" from storage) could get added in the future. At the moment it is
+not needed yet.
+
+The ResourceClaim spec is immutable. The ResourceClaim
+status is reserved for system usage and holds the current state of the
+resource. The status must not get lost, which in the past was not ruled
+out. For example, status could have been stored in a separate etcd instance
+with lower reliability. To recover after a loss, status was meant to be recoverable.
+A [recent KEP](https://github.com/kubernetes/enhancements/tree/master/keps/sig-architecture/2527-clarify-status-observations-vs-rbac)
+clarified that status will always be stored reliably and can be used as
+proposed in this KEP.
+
+To support arbitrarily complex parameters, both ResourceClass and
+ResourceClaim contain one field which holds a
+[runtime.RawExtension](https://pkg.go.dev/k8s.io/apimachinery/pkg/runtime#RawExtension) (see [user stories](#user-stories) for examples).
+Validation can be handled by drivers through an
+admission controller (if desired) or later at runtime when the
+parameters are passed to the driver.
+
+The reasons for choosing embedded RawExtension fields fields instead of
+references to a custom resource (CR) that is defined by a driver's custom
+resource definition (CRD) are:
+
+- Immutability can be enforced: a separate object could always get
+  deleted and recreated with new content.
+- More user-friendly because there is no need to manage an additional
+  object.
+- Less risky resource driver deployment and updates: CRs get deleted when
+  removing their CRD, which would render existing workloads unusable until
+  users redeploy them.
+
+Kube-scheduler
+and resource driver communicate by modifying that status. The status is also
+how Kubernetes tracks that a driver has allocated the resource and on
+which nodes the resource is available.
+
+This approach is an intentional simplification compared to the
+PersistentVolume/PersistentVolumeClaim model for volumes where the additional
+PV object was used to capture status. That model allowed operations like
+pre-provisioning volumes and then having Kubernetes bind those to claims that
+get created later. For resources, the resource driver can and must handle such
+pre-provisioning internally. Kubernetes wouldn't know how to match
```

Comment:
I expect Pod scheduling times to be slower compared to resources handed out by kube-scheduler itself (CPU/RAM) because an external entity needs to be involved. To avoid that, we would have to build (configurable?) vendor logic into kube-scheduler and then keep updating that logic as new hardware adds new constraints.

Reply:
Let's run WebAssembly code in the kube-scheduler for this... not!